intercept/src/intercept.cpp

```diff
@@ -7244,13 +7304,212 @@ void CLIntercept::setKernelArgUSMPointer(
+        return;
+    } 
+    catch (std::exception const& excep)
+    {
+        (void)excep;
+    }
+
+    log("[[Warning]]: Kernel source is not available! Make sure that the kernel is compiled from source (and is not cached)\n");
+    log("Now will try to output binaries, these probably won't work on other platforms!\n");
+    
+    cl_uint num_devices;
+    clGetProgramInfo(tmp_program, CL_PROGRAM_NUM_DEVICES, sizeof(cl_uint), &num_devices, nullptr);
+
+    // Grab the device ids
+    cl_device_id* devices = new cl_device_id[num_devices];
+    clGetProgramInfo(tmp_program, CL_PROGRAM_DEVICES, num_devices * sizeof(cl_device_id), devices, 0);
+
+    std::string knlName = getShortKernelName(kernel);
+    for (unsigned device = 0; device < num_devices; ++device)
+    {
+        auto binaries = m_DeviceBinaryMap[devices[device]];
+        for (unsigned binary = 0; binary != binaries.size(); ++binary)
+        {
+            auto pos = std::search(binaries[binary].data(), binaries[binary].data() + binaries[binary].size(), knlName.data(), knlName.data() + knlName.length());
+            if (pos != binaries[binary].data() + binaries[binary].size())
+            {
+                std::ofstream outputBinaries{fileNamePrefix + "binary_Device_" + std::to_string(device) + "_Binary_" + std::to_string(binary) + ".bin", std::ios_base::binary};
+                outputBinaries.write(reinterpret_cast<const char*>(binaries[binary].data()), binaries[binary].size());
+            }
+        }
+    }
+}
+
+void CLIntercept::dumpKernelInfo(
+    cl_kernel kernel,
+    uint64_t enqueueCounter,
+    size_t work_dim,
+    const size_t* gws_offset,
+    const size_t* gws,
+    const size_t* lws,
+    bool byKernelName)
+{
+    std::lock_guard<std::mutex> lock(m_Mutex);
+    std::string fileNamePrefix = "";
+    OS().GetDumpDirectoryName( sc_DumpDirectoryName, fileNamePrefix );
+    fileNamePrefix += "/Replay/Enqueue_";
+    if (byKernelName)
+        fileNamePrefix += getShortKernelName(kernel);
+    else
+        fileNamePrefix += std::to_string(enqueueCounter);
+    fileNamePrefix += "/";
+    OS().MakeDumpDirectories( fileNamePrefix );
+    std::ofstream output{fileNamePrefix + "worksizes.txt"};
+
+    // Print the values of the worksizes and offsets on a line in the order:
+    // gws
+    // lws
+    // gws_offset
+    for (unsigned idx = 0; idx != work_dim; ++idx)
+    {
+        output << (!gws        ? 0 : gws[idx]) << ' ';
+    }
+    output << '\n';
+
+    for (unsigned idx = 0; idx != work_dim; ++idx)
+    {
+        output << (!lws        ? 0 : lws[idx]) << ' ';
+    }
+    output << '\n';
+
+    for (unsigned idx = 0; idx != work_dim; ++idx)
+    {
+        output << (!gws_offset ? 0 : gws_offset[idx]) << ' ';
+    }
+    output << '\n';
+
+    // Get the cl_program from the cl_kernel and the device(s), then extract kernel build options from the cl_program
+    CLIntercept* pIntercept = GetIntercept();
```

Comment:
Same goes here - we're already in a CLIntercept member function so we don't need to GetIntercept again.

Reply:
Check